cloud-controller-manager/do/certificates_test.go

```diff
@@ -0,0 +1,318 @@
+				lbService.store[lb.ID] = lb
+
+				service, cert := createServiceAndCert(lb.ID, "service-cert-id", certTypeCustom)
+				certService.store[cert.ID] = cert
+				return service
+			},
+			expectedServiceCertID: "service-cert-id",
+			expectedLBCertID:      "service-cert-id",
+		},
+	}
+
+	nodes := []*v1.Node{
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node-1",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node-2",
+			},
+		},
+		{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "node-3",
+			},
+		},
+	}
+	droplets := []godo.Droplet{
+		{
+			ID:   100,
+			Name: "node-1",
+		},
+		{
+			ID:   101,
+			Name: "node-2",
+		},
+		{
+			ID:   102,
+			Name: "node-3",
+		},
+	}
+	fakeDroplet := fakeDropletService{
+		listFunc: func(context.Context, *godo.ListOptions) ([]godo.Droplet, *godo.Response, error) {
+			return droplets, newFakeOKResponse(), nil
+		},
+	}
+
+	for _, tc := range testcases {
+		tc := tc
+		for _, methodName := range []string{"EnsureLoadBalancer", "UpdateLoadBalancer"} {
+			t.Run(tc.name+"_"+methodName, func(t *testing.T) {
+				lbStore := make(map[string]*godo.LoadBalancer)
+				lbService := newKVLBService(lbStore)
+				certStore := make(map[string]*godo.Certificate)
+				certService := newKVCertService(certStore, false)
+				service := tc.setupFn(lbService, certService)
+
+				fakeClient := newFakeClient(&fakeDroplet, &lbService, &certService)
+				fakeResources := newResources("", "", fakeClient)
+				fakeResources.kclient = fake.NewSimpleClientset()
+				if _, err := fakeResources.kclient.CoreV1().Services(service.Namespace).Create(service); err != nil {
+					t.Fatalf("failed to add service to fake client: %s", err)
+				}
+
+				lb := &loadBalancers{
+					resources:         fakeResources,
+					region:            "nyc1",
+					lbActiveTimeout:   2,
+					lbActiveCheckTick: 1,
+				}
+
+				var err error
+				switch methodName {
+				case "EnsureLoadBalancer":
+					_, err = lb.EnsureLoadBalancer(context.TODO(), "test", service, nodes)
+				case "UpdateLoadBalancer":
+					err = lb.UpdateLoadBalancer(context.TODO(), "test", service, nodes)
+				default:
+					t.Errorf("unsupported loadbalancer method: %s", methodName)
+				}
+
+				if !reflect.DeepEqual(err, tc.err) {
+					t.Error("error does not match test case expectation")
+					t.Logf("expected: %v", tc.err)
+					t.Logf("actual: %v", err)
+				}
+
+				service, err = fakeResources.kclient.CoreV1().Services(service.Namespace).Get(service.Name, metav1.GetOptions{})
+				if err != nil {
+					t.Fatalf("failed to get service from fake client: %s", err)
+				}
+
+				serviceCertID := getCertificateID(service)
+				if tc.expectedServiceCertID != serviceCertID {
+					t.Error("unexpected service certificate ID annotation")
+					t.Logf("expected: %s", tc.expectedServiceCertID)
+					t.Logf("actual: %s", serviceCertID)
+				}
+
+				godoLoadBalancer, _, err := lbService.Get(context.TODO(), getLoadBalancerID(service))
```

Comment:
This should use `context.Background()` instead of `context.TODO()` because the latter is supposed to be used

>when it's unclear which Context to use or it is not yet available (because the surrounding function has not yet been extended to accept a Context parameter)

([source](https://golang.org/pkg/context/#TODO))